src/Query/IntermediateTaskGroups.ts

```diff
@@ -11,6 +11,67 @@ import { Group } from './Group';
  */
 export class IntermediateTaskGroupsStorage extends Map<string[], Task[]> {}
 
+/**
+ * A node in the grouping tree. The node contains the
+ * list of tasks matching the path from the root so far, and its children
+ * are the further grouping of those tasks.
+ *
+ * Check the docs of @GroupingTree for more details.
+ *
+ */
+class GroupingTreeNode {
+    children: Map<string, GroupingTreeNode> = new Map();
+    tasks: Task[] = [];
+
+    /**
+     * Recursively traverse the tree to generate all the paths to the leaves.
+     * This function returns a map from every leaf path, to the list of tasks
+     * matching this path.
+     */
+    generateAllPaths(pathSoFar: string[] = []): IntermediateTaskGroupsStorage {
+        const resultMap = new IntermediateTaskGroupsStorage();
+        if (this.children.size == 0) {
+            // Base case: Leaf node. Populate the results map with the path to
+            // this node, and the tasks that match this path.
+            resultMap.set([...pathSoFar], this.tasks);
+            return resultMap;
+        }
+
+        for (const [property, child] of this.children) {
+            pathSoFar.push(property);
+            const childResult = child.generateAllPaths(pathSoFar);
+            childResult.forEach((value, key) => resultMap.set(key, value));
+            pathSoFar.pop();
+        }
+        return resultMap;
+    }
+}
+
+/**
+ * A tree of tasks where every level in the tree corresponds to a grouping property.
+ *
+ * For example, if we have:
+ * # Heading 1
+ * - [ ] Task 1
+ * # Heading 2
+ * - [ ] Task 2
+ * - [X] Task 3
+ *
+ * And we group by heading then status, the tree will look like:
+ *
+ *                   Root [T1, T2, T3]
+ *                     /              \
+ *              Heading 1 [T1]     Heading [T2, T3]
+ *                   |               /        \
+ *               TODO [T1]     TODO [T2]    Done [T3]
+ *
+ * The nice property of this tree is that every path from the root to a leaf, maps
+ * to how the tasks will be rendered.
+ *
+ * NOTE: The same task can appear in multiple leaf nodes, if it matches multiple paths.
+ */
+type GroupingTree = GroupingTreeNode;
```

Comment:
Why are these the same type?

Reply:
I see they aren't now, as the tests use `GroupingTreeNode<number>`. Great stuff!